### PR TITLE
refresh_queries requires Request Context

### PR DESCRIPTION
**Pull request by rauchy**
[Now that `refresh_queries` uses `query.parameterized.apply`](https://github.com/getredash/redash/commit/f0576a362352bd63ae3f96d70bd2d0da1d0646da#diff-5ca0dc869c9c3218d8a95ba7f99f5f4cR196), it requires Flask's Request Context to work.

As this code is used from a background job, it should work without a Request Context. In general we should avoid using Request Context code in models.

### CODE REVIEW

File: redash/tasks/queries.py [hunk 1]
```diff
@@ -190,12 +191,18 @@ def refresh_queries():
...
-                    query_text = query.parameterized.apply(query_params).query
-                else:
-                    query_text = query.query_text
+                query_text = query.query_text
+
+                parameters = {p['name']: p.get('value') for p in query.options.get('parameters', [])}
+                if any(parameters):
+                    try:
+                        query_text = query.parameterized.apply(parameters).query
+                    except InvalidParameterError as e:
```
Comment (rauchy) [inline redash/tasks/queries.py]: I tend to think these should actually not be skipped but rather executed (with an unparameterized `query_text`) and fail - so they will show up in failure reports, in order to give users the chance to rectify them.

**Comment by rauchy** [comment]
Tests coming up.

File: redash/models/__init__.py [hunk 2]
```diff
@@ -262,6 +262,10 @@ def to_dict(self):
             'retrieved_at': self.retrieved_at
         }
 
+    @classmethod
+    def get_by_id(cls, _id):
```
Comment (rauchy) [inline redash/models/__init__.py]: @arikfr how bad is this? What was the motivation behind adding an `org_id` filter on these? The way I understand it, query results identifiers are unique.

Comment (arikfr) [reply redash/models/__init__.py]: It's an extra safety feature to enforce security in a multi tenant context. Even if someone managed to workaround our other safe guards, he will not be able to load a query from a different account. 

TL;DR very bad :)

File: redash/models/parameterized_query.py [hunk 3]
```diff
@@ -17,16 +16,15 @@ def _pluck_name_and_value(default_column, row):
 
 
 def _load_result(query_id):
-    from redash.authentication.org_resolving import current_org
     from redash import models
 
-    query = models.Query.get_by_id_and_org(query_id, current_org)
+    query = models.Query.get_by_id(query_id)
```
Comment (arikfr) [inline redash/models/parameterized_query.py]: Instead of removing the `org_id` argument or relaying on `current_org`, I would pass it to `_load_result`.

Comment (arikfr) [reply redash/models/parameterized_query.py]: I think it should be ok to use the `org_id` value from the Query object we're rendering.

File: redash/tasks/queries.py (hunk 1 above)
Comment (arikfr) [reply redash/tasks/queries.py]: We do the reporting at a different stage and hence why we can't report on them. If we move the parameter merging to the query task, then we can report on this issue.

File: redash/models/parameterized_query.py (hunk 3 above)
Comment (rauchy) [reply redash/models/parameterized_query.py]: > Instead of removing the org_id argument or relaying on current_org, I would pass it to _load_result.

Figured we'd end up with that but thought I'd see if it's avoidable first.

File: redash/tasks/queries.py (hunk 1 above)
Comment (rauchy) [reply redash/tasks/queries.py]: Heh, I was actually thinking of a far less elegant solution - when unable to do parameter merging - just let a broken unapplied query run and have the DB report a failure, which would end up in the report :)

File: redash/models/__init__.py (hunk 2 above)
Comment (rauchy) [reply redash/models/__init__.py]: 9cfadd8 😖

File: redash/models/parameterized_query.py (hunk 3 above)
Comment (rauchy) [reply redash/models/parameterized_query.py]: Some `ParameterizedQuery` instances do not represent a persisted `Query` so we can't rely on having it all the time. Instead I just pass in the actual org 😖

File: redash/tasks/queries.py (hunk 1 above)
Comment (rauchy) [reply redash/tasks/queries.py]: One caveat about moving the parameter merging to the query task - `enqueue_query` creates job locks according to the hashed query. If we don't apply the parameters before calling `enqueue_query`, we will lock according to hashed versions of strings like `SELECT {{foo}}`, which could prevent different queries from running if they hash to the same value.

This is probably a non-issue because scheduled parameterized queries use their default values, so they hash to the same value regardless. It's just disturbing to see.

Comment (arikfr) [reply redash/tasks/queries.py]: We should revise the hashing logic anyway, as it's currently lower cases the query. We can hash the query text + parameter values to de-duplicate based on actual query without merging in the values.

File: redash/tasks/queries.py [hunk 4]
```diff
@@ -190,12 +191,20 @@ def refresh_queries():
...
             else:
-                if query.options and len(query.options.get('parameters', [])) > 0:
-                    query_params = {p['name']: p.get('value')
-                                    for p in query.options['parameters']}
-                    query_text = query.parameterized.apply(query_params).query
-                else:
-                    query_text = query.query_text
+                query_text = query.query_text
+
+                parameters = {p['name']: p.get('value') for p in query.options.get('parameters', [])}
```
Comment (arikfr) [inline redash/tasks/queries.py]:
```suggestion
                parameters = {p['name']: p.get('value') for p in query.parameters}
```

File: redash/tasks/queries.py (hunk 1 above)
Comment (arikfr) [reply redash/tasks/queries.py]: Another option is to report from here directly to the error report. We should also increment the schedule failures counter to avoid having to reprocess these queries all the time.

Out of scope for this PR, but worth keeping in mind.

File: redash/tasks/queries.py [hunk 5]
```diff
@@ -190,12 +191,20 @@ def refresh_queries():
...
+                if any(parameters):
+                    try:
+                        query_text = query.parameterized.apply(parameters).query
+                    except InvalidParameterError as e:
+                        logging.info("Skipping refresh of %s because of invalid parameters: %s", query.id, e.message)
```
Comment (arikfr) [inline redash/tasks/queries.py]: This and the other log message have the potential to significantly spam the log (in a busy system). Let's move them to `debug` for now.

File: redash/handlers/queries.py [hunk 6]
```diff
@@ -451,7 +451,7 @@ def post(self, query_id):
         require_access(query, self.current_user, not_view_only)
 
         parameter_values = collect_parameters_from_request(request.args)
-        parameterized_query = ParameterizedQuery(query.query_text)
+        parameterized_query = ParameterizedQuery(query.query_text, org=self.current_org)
```
Comment (arikfr) [inline redash/handlers/queries.py]: Why aren't we using query.paremterized_query here?

Comment (rauchy) [reply redash/handlers/queries.py]: Using `query.parameterized` would create a `ParameterizedQuery` with the query schema baked into it which would, in turn, force parameter schema validation.

It might be desirable to verify the schema here as well, I just didn't want to change any behavior in that part of the code as part of this PR. Let me know if you think it's safe.

Comment (arikfr) [reply redash/handlers/queries.py]: Got it. Keep as is.

File: redash/tasks/queries.py (hunk 1 above)
Comment (rauchy) [reply redash/tasks/queries.py]: Went ahead with the latter. Let me know if you want to keep cb82a53 or discard it.

File: redash/tasks/queries.py [hunk 7]
```diff
@@ -366,10 +374,7 @@ def run(self):
             result = QueryExecutionError(error)
             if self.scheduled_query is not None:
                 self.scheduled_query = models.db.session.merge(self.scheduled_query, load=False)
-                self.scheduled_query.schedule_failures += 1
-                models.db.session.add(self.scheduled_query)
-                notify_of_failure(error, self.scheduled_query)
-            models.db.session.commit()
+                track_failure(self.scheduled_query, error)
```
Comment (arikfr) [inline redash/tasks/queries.py]: Previously, `models.db.session.commit` was happening in all cases. Now it happens only on failure.

Comment (rauchy) [reply redash/tasks/queries.py]: First - I think you mean it was happening for all queries and now it only happens with scheduled queries?

Second - I have a hunch this was a mistake with indentation, as there is no `db.session` work prior to it.

Comment (arikfr) [reply redash/tasks/queries.py]: I thought it's for saving the result, but that actually happens in a different branch of the code. So I guess you're right.

File: redash/tasks/queries.py [hunk 8]
```diff
@@ -182,20 +183,28 @@ def refresh_queries():
     with statsd_client.timer('manager.outdated_queries_lookup'):
         for query in models.Query.outdated_queries():
             if settings.FEATURE_DISABLE_REFRESH_QUERIES:
-                logging.info("Disabled refresh queries.")
+                logging.debug("Disabled refresh queries.")
```
Comment (arikfr) [inline redash/tasks/queries.py]: This shouldn't be a debug message, because it low frequency and important.

```suggestion
                logging.info("Disabled refresh queries.")
```

**Comment by arikfr** [comment]
:+1:
